File: kmp/Pulsar/library/src/iosMain/kotlin/com/swmansion/pulsar/Lines/CurveLineModifier.kt
package com.swmansion.pulsar.kmp

import kotlinx.cinterop.ExperimentalForeignApi
import platform.CoreHaptics.CHHapticParameterCurve
import platform.CoreHaptics.CHHapticParameterCurveControlPoint

@OptIn(ExperimentalForeignApi::class)
internal open class IOSCurveLineModifier(
    private val parameterId: String?,
) {
    private val points = mutableListOf<CHHapticParameterCurveControlPoint>()
    private var maxTime = 0.0

    val isEmpty: Boolean
        get() = points.isEmpty()

    fun addPoint(time: Long, value: Float) {
        val timeInSeconds = time.toDouble() / 1000.0
        if (points.isEmpty() && timeInSeconds > 0.0) {
            points += CHHapticParameterCurveControlPoint(relativeTime = 0.0, value = value)
        }
        points += CHHapticParameterCurveControlPoint(relativeTime = timeInSeconds, value = value)
        if (timeInSeconds > maxTime) {
            maxTime = timeInSeconds
        }
    }

    fun getDuration(): Double = maxTime

    fun reset() {
        points.clear()
        maxTime = 0.0
    }

    val curve: CHHapticParameterCurve
        get() = CHHapticParameterCurve(parameterID = parameterId, controlPoints = points, relativeTime = 0.0)
}

File: flutter/pulsar/ios/Classes/Lines/CurveLineModifier.swift
import CoreHaptics


public class CurveLineModifier {
  var points: [CHHapticParameterCurve.ControlPoint] = []
  var maxTime: Double = 0
  
  /// Adds a control point to the curve
  /// - Parameters:
  ///   - time: Time in milliseconds (ms)
  ///   - value: Normalized value (0.0 - 1.0)
  public func addPoint(time: Double, value: Float) {
    let timeInSeconds = time / 1000.0
    if points.isEmpty && timeInSeconds > 0 {
      points.append(
        CHHapticParameterCurve.ControlPoint(relativeTime: 0, value: value)
      )
    }
    points.append(
      CHHapticParameterCurve.ControlPoint(relativeTime: timeInSeconds, value: value)
    )
    if timeInSeconds > maxTime {
      maxTime = timeInSeconds
    }
  }
  
  public func getDuration() -> Double {
    return maxTime
  }
  
  public func reset() {
    points.removeAll()
    maxTime = 0
  }
  
  public var isEmpty: Bool {
    return points.isEmpty
  }
}
